File: c++/Day8.cpp
#include "tools.h"
#include <tl/getlines.hpp>
#include <tl/to.hpp>
#include <print>
#include <iostream>
#include <fstream>

int main() {
    std::ifstream file ("C:/Users/navez/Cpp_Projects/AoC/Advent_2022_Files/Day8.txt");
    auto s = tl::views::getlines(file) | tl::to<std::vector<std::string>>();

    int count = (s.size() * 2 + s[0].size() * 2) - 4;
    vi scores;

    for (size_t rows = 1; rows < s.size() - 1; rows++) {
        for (size_t cols = 1; cols < s[cols].size() - 1; cols++) {
            int pos = (int) s[rows][cols];
            vi left, right, top, bottom;

            for (size_t i = 1; i < cols + 1; i++) left.emplace_back((int) s[rows][cols - i]);
            for (size_t i = 1; i < s.size() - cols; i++) right.emplace_back((int) s[rows][cols + i]);
            for (size_t i = 1; i < rows + 1; i++) top.emplace_back((int) s[rows - i][cols]);
            for (size_t i = 1; i < s.size() - rows; i++) bottom.emplace_back((int) s[rows + i][cols]);

            if (maxVal(left) < pos || maxVal(right) < pos || maxVal(top) < pos || maxVal(bottom) < pos) count++;

            int score {1};
            for (const auto &lst: {left, right, top ,bottom}) {
                int tracker {0};
                for (size_t i {0}; i < lst.size(); i++) {
                    if (lst[i] < pos) tracker++;
                    else if (lst[i] == pos) {
                        tracker++;
                        break;
                    }
                    else break;
                }
                score *= tracker;
                scores.emplace_back(score);
            }
        }
    }
    
    std::print("Part 1: {}\nPart 2: {}\n", count, *std::max_element(scores.begin(), scores.end()));
}

File: nim/Day8.nim
import strformat

proc readInput(): seq[string] =
    var s: seq[string]
    let file = open("input/Day8.txt");
    defer: file.close()

    for i in file.lines():
        s.add(i)
    
    return s
        
block:
    let s = readInput()
    var
        count = (len(s) * 4) - 4
        scores: seq[int]

    for rows in 1..<len(s) - 1:
        for cols in 1..<len(s) - 1:
            var pos = s[rows][cols]

            var
                left: seq[char]
                right: seq[char]
                top: seq[char]
                bottom: seq[char] 

            for i in 1..<cols + 1:      left.add((s[rows][cols - i]))
            for i in 1..<len(s) - cols: right.add((s[rows][cols + i]))
            for i in 1..<rows + 1:      top.add((s[rows - i][cols]))
            for i in 1..<len(s) - rows: bottom.add((s[rows + i][cols]))

            if max(left) < pos or max(right) < pos or max(top) < pos or max(bottom) < pos:
                count += 1

            var score = 1
            for lst in [left, right, top, bottom]:
                var tracker = 0
                for i in 0..<len(lst):
                    if lst[i] < pos:
                        tracker += 1
                    elif lst[i] == pos:
                        tracker += 1
                        break
                    else:
                        break
            
                score *= tracker
                scores.add(score)
    
    echo &"Solution 1: {count}\nSolution 2: {max(scores)}"
